File: src/yx/transport/protocol_router.py
"""
Protocol Router - Dispatches packets to protocol handlers.

Traceability:
- protocol/specs/architecture/protocol-layers.md (Protocol Router section)
"""

from enum import IntEnum
from typing import Dict, Callable, Awaitable, Optional
import logging

logger = logging.getLogger(__name__)


class ProtocolID(IntEnum):
    """
    Protocol identifier byte (first byte of payload).

    Traceability:
    - protocol/specs/architecture/protocol-layers.md (Protocol ID Registry)
    """
    TEXT = 0x00     # Protocol 0: Text/JSON-RPC
    BINARY = 0x01   # Protocol 1: Binary/Chunked


class ProtocolRouter:
    """
    Routes validated payloads to appropriate protocol handler.

    Traceability:
    - protocol/specs/architecture/protocol-layers.md (Protocol Router)
    """

    def __init__(self):
        self._handlers: Dict[int, Callable[[bytes], Awaitable[None]]] = {}

    def register(
        self,
        protocol_id: int,
        handler: Callable[[bytes], Awaitable[None]]
    ):
        """
        Register protocol handler.

        Args:
            protocol_id: Protocol ID byte (e.g., 0x00 for text)
            handler: Async function that processes payload
        """
        self._handlers[protocol_id] = handler
        logger.info(f"Registered protocol handler for ID: 0x{protocol_id:02x}")

    async def route(self, payload: bytes):
        """
        Route payload to handler by protocol ID.

        Args:
            payload: Raw payload (first byte is protocol ID)
        """
        if not payload:
            logger.debug("Empty payload, ignoring")
            return

        protocol_id = payload[0]

        handler = self._handlers.get(protocol_id)
        if handler is None:
            logger.error(f"Unknown protocol ID: 0x{protocol_id:02x}")
            return

        try:
            await handler(payload)
        except Exception as e:
            logger.exception(f"Error in protocol handler 0x{protocol_id:02x}: {e}")

File: canonical/swift/Sources/YXProtocol/Transport/ProtocolRouter.swift
import Foundation

/// Routes incoming payloads to protocol-specific handlers based on protocol ID
///
/// Protocol IDs:
/// - 0x00: Text (JSON-RPC 2.0)
/// - 0x01: Binary (Chunked with compression/encryption)
actor ProtocolRouter {

    /// Protocol handler function type
    typealias ProtocolHandler = (Data) async throws -> Void

    /// Registered protocol handlers
    private var handlers: [UInt8: ProtocolHandler] = [:]

    /// Register a protocol handler
    func register(protocolID: UInt8, handler: @escaping ProtocolHandler) {
        handlers[protocolID] = handler
    }

    /// Route payload to appropriate protocol handler
    /// - Throws: ProtocolError if protocol not supported or payload invalid
    func route(payload: Data) async throws {
        guard !payload.isEmpty else {
            throw ProtocolError.emptyPayload
        }

        let protocolID = payload[payload.startIndex]

        guard let handler = handlers[protocolID] else {
            throw ProtocolError.unsupportedProtocol(protocolID)
        }

        try await handler(payload)
    }

    /// Get registered protocol IDs
    func registeredProtocols() -> [UInt8] {
        return Array(handlers.keys).sorted()
    }
}

/// Protocol routing errors
enum ProtocolError: Error, CustomStringConvertible {
    case emptyPayload
    case unsupportedProtocol(UInt8)
    case invalidFormat(String)
    case encodingError(String)
    case decodingError(String)

    var description: String {
        switch self {
        case .emptyPayload:
            return "Empty payload"
        case .unsupportedProtocol(let id):
            return "Unsupported protocol: 0x\(String(format: "%02X", id))"
        case .invalidFormat(let msg):
            return "Invalid format: \(msg)"
        case .encodingError(let msg):
            return "Encoding error: \(msg)"
        case .decodingError(let msg):
            return "Decoding error: \(msg)"
        }
    }
}
